File: leetcode/easy/94-inorder/morris.py
"""
  @author Vortexx2
  Problem 94 - Binary Tree Inorder Traversal
  
  Runtime - 32 ms
  Memory Usage - 14.4 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def inorderTraversal(self, root: TreeNode) -> List[int]:
    res = []

    while root:

      if root.left:
        predec = self.findPredecessor(root)

        if predec.right:
          predec.right = None
          res.append(root.val)
          root = root.right

        else:
          predec.right = root
          root = root.left

      else:
        res.append(root.val)
        root = root.right

    return res

  def findPredecessor(self, node: TreeNode) -> TreeNode:
    curr = node.left
    while (curr.right and curr.right != node):
      curr = curr.right

    return curr

File: leetcode/easy/94-inorder/morris.cpp
/**
 * @file morris.cpp
 * @author Vortexx2
 * @brief Problem 94 - Binary Tree Inorder Traversal
 * @date 22-07-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 8.2 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  /**
   * @brief Using the *Morris Traversal Algorithm* to get a solution which is O(n) in time and O(1) in space complexity.
   * 
   * @param root 
   * @return vector<int> 
   */
  vector<int> inorderTraversal(TreeNode *root)
  {
    vector<int> res;

    while (root)
    {
      if (root->left)
      {
        TreeNode *predecessor = findPredecessor(root);
        if (!predecessor->right)
        {
          predecessor->right = root;
          root = root->left;
        }

        else
        {
          predecessor->right = NULL;
          res.push_back(root->val);
          root = root->right;
        }
      }
      else
      {
        res.push_back(root->val);
        root = root->right;
      }
    }

    return res;
  }

  /**
   * @brief Finds the predecessor for a particular node.
   * 
   * @param node 
   * @return TreeNode* 
   */
  TreeNode *findPredecessor(TreeNode *node)
  {
    TreeNode *curr = node;
    // node->left will exist since we only call method when left exists
    curr = curr->left;

    while (curr->right && curr->right != node)
      curr = curr->right;

    return curr;
  }
};